presto-manager-controller/src/main/java/com/teradata/prestomanager/controller/NodeSet.java

```diff
@@ -0,0 +1,123 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.teradata.prestomanager.controller;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import java.net.URI;
+import java.util.Collection;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public class NodeSet
+{
+    private static NodeSet nodeSetInstance = new NodeSet();
+
+    public static Set<Agent> nodeSet = ConcurrentHashMap.newKeySet();
+
+    public static NodeSet getInstance()
+    {
+        return nodeSetInstance;
+    }
+
+    public void addAgent(URI uri, boolean isCoordinator, boolean isWorker)
+    {
+        nodeSet.add(new Agent(uri, isCoordinator, isWorker));
+    }
+
+    public void removeAgent(int id)
+    {
+        nodeSet.removeIf(agent -> agent.getId().equals(id));
+    }
+
+    public Collection<URI> getUrisByIds(Collection<Integer> ids)
+    {
+        Collection<URI> uriList = nodeSet.stream()
+                .filter(agent -> ids.contains(agent.getId()))
+                .map(Agent::getUri).collect(Collectors.toList());
+        if (ids.size() != uriList.size()) {
+            throw new IllegalArgumentException("Invalid or duplicate node ID");
+        }
+        else {
+            return uriList;
+        }
+    }
+
+    public Collection<URI> getUris(ApiScope scope)
+    {
+        switch (scope) {
+            case CLUSTER:
+                return nodeSet.stream().map(Agent::getUri).collect(Collectors.toList());
+            case COORDINATOR:
+                return nodeSet.stream().filter(Agent::isCoordinator)
+                        .map(Agent::getUri).collect(Collectors.toList());
+            case WORKERS:
+                return nodeSet.stream().filter(Agent::isWorker)
+                        .map(Agent::getUri).collect(Collectors.toList());
+            default:
+                return null;
+        }
+    }
+
+    //TODO: Add method to generate ID
+    public static class Agent
+    {
+        private URI uri;
+        private static final AtomicInteger count = new AtomicInteger(0);
+        private int id;
+        private boolean isCoordinator;
+        private boolean isWorker;
+
+        private Agent(URI uri, boolean isCoordinator, boolean isWorker)
+        {
+            this.uri = requireNonNull(uri, "uri is null");
+            this.isCoordinator = isCoordinator;
+            this.isWorker = isWorker;
+
+            this.id = count.incrementAndGet();
+        }
+
+        private URI getUri()
+        {
+            return uri;
+        }
+
+        private Integer getId()
+        {
+            return id;
+        }
+
+        private boolean isCoordinator() {return isCoordinator;}
```

Comment:
Add newlines here, like the other getters -- the Presto style guide checks for this, you might want to add that.